src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java

```diff
@@ -464,16 +607,58 @@ private double getElevation(double x, double y) {
             // for DefaultGeographicCRS.WGS84, but OTP is using (long, lat) throughout and assumes unprojected DEM
             // rasters to also use (long, lat).
             coverage.evaluate(new DirectPosition2D(GeometryUtils.WGS84_XY, x, y), values);
-        } catch (org.opengis.coverage.PointOutsideCoverageException e) {
-            nPointsOutsideDEM += 1;
+        } catch (PointOutsideCoverageException e) {
+            nPointsOutsideDEM.incrementAndGet();
+            throw e;
         }
-        nPointsEvaluated += 1;
-        return values[0];
+        nPointsEvaluated.incrementAndGet();
+        return values[0] - (includeEllipsoidToGeoidDifference ? getApproximateEllipsoidToGeoidDifference(y, x) : 0);
+    }
+
+    /**
+     * The Calculation of the EllipsoidToGeoidDifference is a very expensive operation, so the resulting values are
+     * cached based on the coordinate values up to 2 significant digits. Two signficant digits is often more than enough
+     * for most parts of the world, but is useful for certain areas that have dramatic changes. Since the values are
+     * computer once and cached, it has almost no affect on performance to have this level of detail.
+     * See this image for an approximate mapping of these difference values:
+     * https://earth-info.nga.mil/GandG/images/ww15mgh2.gif
+     *
+     * @param y latitude
+     * @param x longitude
+     */
+    private double getApproximateEllipsoidToGeoidDifference(double y, double x) throws TransformException {
+        int geoidDifferenceCoordinateValueMultiplier = 100;
+        int xVal = (int) Math.round(x * geoidDifferenceCoordinateValueMultiplier);
+        int yVal = (int) Math.round(y * geoidDifferenceCoordinateValueMultiplier);
+        // create a hash value that can be used to look up the value for the given rounded coordinate. The expected
+        // value of xVal should never be less than -18000 (-180 * 100) or more than 18000 (180 * 100), so multiply the
+        // yVal by a prime number of a magnitude larger so that there won't be any hash collisions.
+        int hash = yVal * 104729 + xVal;
+        Double difference = geoidDifferenceCache.get(hash);
+        if (difference == null) {
+            difference = computeEllipsoidToGeoidDifference(
+                yVal / (1.0 * geoidDifferenceCoordinateValueMultiplier),
```

Comment:
Can't you use x and y that are already passed in the arguments?

Reply:
Ok I see why, to make the {hash, diff} consistent regardless of the original x and y. I thought that was a wasted conversion back to double.

Reply:
The rounded values are used because this way all coordinates that get rounded will use the same consistent coordinate as opposed to whatever exact coordinate values are in the arguments.